control-plane/p4RuntimeSerializer.cpp

```diff
@@ -639,6 +640,45 @@ getMatchType(cstring matchTypeName) {
     }
 }
 
+static int
+getTypeWidth(const IR::Type* type, TypeMap* typeMap) {
+    // p4runtime tests use bit Slices which are Type_Bits
+    // which will cause failure in minWidthBits.
+    // Thus compute width here for Slices and any other bit-field.
+    if (auto tb = type->to<IR::Type_Bits>()) {
```

Comment:
@mbudiu-vmw This sounds like an issue with `minWidthBits` or the type map, and not something for which we should have a workaround here. Mihai, do you know what could be happening. If I understand the issue correctly, when a match field is a Slice expression, and we call this function (`getTypeWidth`) on the match field's type (which is `Type_Bits`), the compiler crashes (I'm sure Hemant can provide a backtrace). Things work fine when the match field is just a plain field path.

Reply:
If a Slice's type is sent to `getTypeType`, it calls `getType` which ends up into this `BUG` condition.

https://github.com/p4lang/p4c/blob/master/frontends/p4/typeMap.cpp#L84

The reason is because the `element` is not found in the typeMap.  `getTypeType` is called by `minWidthBits` and also my new code in p4runtime serializer.  The reason the type is not found in the typeMap is because any caller to minWidthBits or getTypeName is only getting the type for the Slice.  The actual Slice expression is `h.h.headerField[13:4]` which exists in typeMap.

The Slice IR dump from getMatchFields is included below when gtestp4c is run.

```
[1223064] KeyElement
  annotations: [1145486] Annotations
    [1145488] Annotation name=name needsParsing=0
      [1145492] StringLiteral value=h.h.headerField[13:4]
        type: [11866] Type_String
  expression: [1223072] Slice
    type: [959116] Type_Bits size=10 isSigned=0
    e0: [1223073] Member member=headerField
      type: [6419] Type_Bits size=16 isSigned=0
      expr: [1223074] Member member=h
        type: [1119732] Type_Header name=Header declid=3516
          annotations: [2] Annotations
          [1119729] StructField name=headerField declid=9657
            annotations: [2] Annotations
            type: [6419] Type_Bits size=16 isSigned=0...
        expr: [1223075] PathExpression
          type: [1217347] Type_Struct name=Headers declid=3703
            annotations: [2] Annotations
            [1217342] StructField name=h declid=10176
              annotations: [2] Annotations
              type: [1119732] Type_Header name=Header declid=3516...
            [1217343] StructField name=hStack declid=10177
              annotations: [2] Annotations
              type: [1208349] Type_Stack
                elementType: [1119732] Type_Header name=Header declid=3516...
                size: [1119771] Constant value=4 base=10
                  type: [1119770] Type_InfInt declid=1162
            [1217344] StructField name=hUnion declid=10178
              annotations: [2] Annotations
              type: [1217284] Type_HeaderUnion name=HeaderUnion declid=3702
                annotations: [2] Annotations
                [1217279] StructField name=a declid=10170
                  annotations: [2] Annotations
                  type: [1119732] Type_Header name=Header declid=3516...
                [1217280] StructField name=b declid=10171
                  annotations: [2] Annotations
                  type: [1119740] Type_Header name=AnotherHeader declid=3517
                    annotations: [2] Annotations
                    [1119737] StructField name=anotherHeaderField declid=9658
                      annotations: [2] Annotations
                      type: [190009] Type_Bits size=8 isSigned=0
                [1217281] StructField name=c declid=10172
                  annotations: [2] Annotations
                  type: [1217245] Type_Header name=YetAnotherHeader declid=3701
                    annotations: [2] Annotations
                    [1217242] StructField name=yetAnotherHeaderField declid=10168
                      annotations: [2] Annotations
                      type: [1119725] Type_Newtype name=CustomT_t declid=3515
                        annotations: [1119723] Annotations
                        type: [190009] Type_Bits size=8 isSigned=0...
          path: [1120128] Path name=h absolute=0
    e1: [1120132] Constant value=13 base=10
      type: [1120131] Type_InfInt declid=1174
    e2: [1120134] Constant value=4 base=10
      type: [1120133] Type_InfInt declid=1175
  matchType: [1145502] PathExpression
    type: [11847] Type_MatchKind
    path: [1120137] Path name=ternary absolute=0
```

Reply:
The function getTypeType is being called with a type that has not been inserted in the typeMap. Typechecking inserts all types that appear in the IR in the typemap, so some other pass has created this type node without inserting it into the typeMap. The way this is debugged is by setting a conditional breakpoint on the creation of the id of the IR::Node, which is unique.

Reply:
If the type is not expected to be in the typeMap maybe we should not use typeMap->getMinWidth anyway. The function to compute the width should be very simple for the types supported by the control-plane.

Reply:
> If the type is not expected to be in the typeMap maybe we should not use typeMap->getMinWidth anyway. The function to compute the width should be very simple for the types supported by the control-plane.

This is precisely what I have done in this PR at this line of code and avoided calling getMinWidth.

https://github.com/p4lang/p4c/blob/05e7aba7c6f597e2ccc9303db8af7ad5dc5b58cd/control-plane/p4RuntimeSerializer.cpp#L649

Reply:
> The function getTypeType is being called with a type that has not been inserted in the typeMap. Typechecking inserts all types that appear in the IR in the typemap, so some other pass has created this type node without inserting it into the typeMap. The way this is debugged is by setting a conditional breakpoint on the creation of the id of the IR::Node, which is unique.

There is nothing to debug here and no other pass is involved.  In p4runtime Serializer,  the function in getMatchFields calls getTypeName and getTypeWidth for a table's MatchFields.  When the MatchField is a bit Slice, the 'matchFieldType` variable in the function is set to Type_Bits.  This is why, as discussed in another comment, getTypeWidth does not call minWidthBits and returns tb->width_bits().  In the same way, getTypeName has been changed so that if incoming type is Type_Bits, bail out because (a) the subsequent getTypeType will fail with a BUG, and (b) if type is Type_Bits, a type_name is nullptr for such a type - this behavior is by design.  In summary, the code change is fine.

Reply:
I agree with Mihai (at least with his first comment). There *is* something to debug, unless you have a good explanation for why it is not in the typeMap. So far all I see from you is a description of the problem, not the explanation.
I'm not saying it is a problem with your code, but I believe that if something is not in the typeMap, there is an issue somewhere. We do run the type checker in p4RuntimeSerializer (https://github.com/p4lang/p4c/blob/master/control-plane/p4RuntimeSerializer.cpp#L1532) after a few minor passes, so it seems that the typeMap should be up-to-date.

Reply:
If the Slice is not in typeMap, why didn't this line of code get invoked to cause an error?

https://github.com/p4lang/p4c/blob/master/control-plane/p4RuntimeSerializer.cpp#L662

Reply:
> The function getTypeType is being called with a type that has not been inserted in the typeMap. Typechecking inserts all types that appear in the IR in the typemap, so some other pass has created this type node without inserting it into the typeMap. The way this is debugged is by setting a conditional breakpoint on the creation of the id of the IR::Node, which is unique.

@mbudiu-vmw Could you please give an example for how to set the breakpoint in gdb.  I don't know the id for IR::Node and what is the creation function.

Reply:
@mbudiu-vmw All bit-field MatchFields in p4c/test/getest/p4runtime.cpp that succeed with getTypeType() have this IR as being Type_Type.  I obtained this information by debugging getType() in typeMap.cpp.

```
[1214926] Type_Type
  type: [6419] Type_Bits size=16 isSigned=0
```

The failing bit Slice MatchField has this IR

```
[959116] Type_Bits size=10 isSigned=0
```

This is why getTypeType() fails for Slice.  I think, in typeChecker or elsewhere where a table key is being added to typeMap, the Slice has to be marked as Type_Type for top-level type.   Please see this line of p4runtime.cpp above which several bit-fields exist which all have Type_Type.  

https://github.com/p4lang/p4c/blob/master/test/gtest/p4runtime.cpp#L428

If the Slice IR cannot be set to Type_Type, then we do have root cause and have to avoid the issue in p4runtimeSerializer code.

Reply:
> According to your own comments there is something to debug: a program for which a type is not present in the typeMap.

I did debug as far as I could - I don't know the unique ID for IR::Node to debug its constructor.  I also debugged postorder(IR::Slice*) in typeChecker and didn't see anything obviously awry.  For good cases, getType() returns Type_Type as result.  I have added a log in another note below.

Reply:
One thought: if you try generating P4Info for a P4_16 program with a slice expression in a match key, do you observe the same issue? If not, maybe it is an issue with how the GTest test case is crafted and we can look into it.

Reply:
Good thought, Antonin.  I prepared a small P4 program with a table key as Slice expression and see the same failure as Gtest.  I will add the test program to the issue I filed last night.  thanks.

```
hemant@ubuntu:/tmp/p4c/build$ ./p4test --std p4-16 ../testdata/p4_16_samples/foo.p4 --p4runtime-files tmp.txt
terminate called after throwing an instance of 'Util::CompilerBug'
  what():  In file: /tmp/p4c/frontends/p4/typeMap.cpp:84
Compiler Bug: Could not find type for <Type_Bits>(1161) bit<10>

hemant@ubuntu:/tmp/p4c/build$ 
```